lib/rubocop/cop/rspec/verified_double_reference.rb

```diff
@@ -0,0 +1,119 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module RSpec
+      # Checks for consistent verified double reference style.
+      # Only investigates references that are one of the supported styles.
+      #
+      # @see https://relishapp.com/rspec/rspec-mocks/docs/verifying-doubles
+      #
+      # This cop can be configured in your configuration using the
+      # `EnforcedStyle` option and supports `--auto-gen-config`.
+      #
+      # @example `EnforcedStyle: constant`
+      #   # bad
+      #   let(:foo) do
+      #     instance_double('ClassName', method_name: 'returned_value')
+      #   end
+      #
+      #   # good
+      #   let(:foo) do
+      #     instance_double(ClassName, method_name: 'returned_value')
+      #   end
+      #
+      # @example `EnforcedStyle: string`
+      #   # bad
+      #   let(:foo) do
+      #     instance_double(ClassName, method_name: 'returned_value')
+      #   end
+      #
+      #   # good
+      #   let(:foo) do
+      #     instance_double('ClassName', method_name: 'returned_value')
+      #   end
+      #
+      # @example Reference is not in the supported style list. No enforcement
+      #
+      #   # good
+      #   let(:foo) do
+      #     instance_double(@klass, method_name: 'returned_value')
+      #   end
+      class VerifiedDoubleReference < Base
+        extend AutoCorrector
+        include ConfigurableEnforcedStyle
+
+        MSG = 'Prefer `%<style>s` references for verified doubles.'
+
+        RESTRICT_ON_SEND = Set[
+          :class_double,
+          :class_spy,
+          :instance_double,
+          :instance_spy,
+          :mock_model,
+          :object_double,
+          :object_spy,
+          :stub_model
+        ].freeze
+
+        NODE_STYLE_MAP = {
+          str: :string,
+          const: :constant
+        }.freeze
+
+        # @!method verified_double(node)
+        def_node_matcher :verified_double, <<~PATTERN
+          (send
+            nil?
+            %RESTRICT_ON_SEND # unpacks constant
+            $_ # class reference to inspect
```

Comment:
Cosmetic: can be written as `$_class_reference` if you want to make it clear what it captures exactly.